File: Python/_28_Find_the_Index_of_the_First_Occurrence_in_a_String.py
class Solution:
    """
    Given two strings `needle` and `haystack`, return the index of the first occurrence of `needle` in `haystack`, or `-1` if `needle` is not part of `haystack`.

    #### Example 1:
    **Input:** haystack = "sadbutsad", needle = "sad" \\
    **Output:** 0 \\
    **Explanation:** "sad" occurs at index 0 and 6. The first occurrence is at index 0, so we return 0. \
    
    #### Example 2:
    **Input:** haystack = "leetcode", needle = "leeto" \\
    **Output:** -1 \\
    **Explanation:** "leeto" did not occur in "leetcode", so we return -1.

    #### Constraints:
    - `1 <= haystack.length, needle.length <= 10^4`
    - `haystack` and `needle` consist of only lowercase English characters.
    """

    def strStr(self, haystack: str, needle: str) -> int:
        """
        Return the index of the first occurrence of `needle` in `haystack`, or `-1` if `needle` is not part of `haystack`.

        Args:
            haystack (str): the string
            needle (str): the substring for which to search

        Returns:
            int: the index of the first occurence of the substring
        """
        return haystack.find(needle)


def main():
    sol = Solution()
    print(sol.strStr("leetcode", "leeto"))

if __name__ == "__main__":
    main()

File: Java/_28_Find_the_Index_of_the_First_Occurrence_in_a_String.java

public class _28_Find_the_Index_of_the_First_Occurrence_in_a_String {

    /**
     * Given two strings {@code needle} and {@code haystack}, return the index
     * of the first occurrence of {@code needle} in {@code haystack}, or
     * {@code -1} if {@code needle} is not part of {@code haystack}
     * <pre>
     *   <b>Example 1:</b>
     *   <b>Input:</b> haystack = "sadbutsad", needle = "sad"
     *   <b>Output:</b> 0
     *   <b>Explanation:</b> "sad" occurs at index 0 and 6. The first occurrence is at index 0, so we return 0.
     * 
     *   <b>Example 2:</b>
     *   <b>Input:</b> haystack = "leetcode", needle = "leeto"
     *   <b>Output:</b> -1
     *   <b>Explanation:</b> "leeto" did not occur in "leetcode", so we return -1.
     * </pre>
     * <b>Constraints:</b>
     * <ul>
     * <li>{@code 1 <= haystack.length, needle.length <= 10^4}</li>
     * <li>{@code haystack} and {@code needle} consist of only lowercase English
     * characters.</li>
     * </ul>
     */
    static class Solution {

        /**
         * Returns the index within this string of the first occurrence of the specified substring.
         * 
         * @param haystack the string
         * @param needle the substing
         * @return the index of the first occurrence of the specified substring, or -1 if there is no such occurrence
         */
        public int strStr(String haystack, String needle) {
            return haystack.indexOf(needle);
        }
    }

    public static void main(String[] args) {
        var sol = new Solution();
        System.out.println(sol.strStr("stringTesting", "Testing"));
    }
}
